### PR TITLE
feat: dynamic interface implementing for ResponseWriter

### LINKED ISSUE #265: negroni.ResponseWriter implements http.Flusher even if underlying http.ResponseWriter doesn't
negroni.ResponseWriter wrapper of http.ResponseWriter always implements all of http ResponseWriter extensions like http.Flusher, http.Hijacker and etc. But implementation is really dummy inside and doesn't provide guarantees given by these interfaces. For example Flush() method just checks underlying ResponseWriter and calls it's Flush() method.

That may cause problems with the code, that rely it's logic based on the support of such interfaces, like:

```go
func (w http.ResponseWriter) {
    if wf, ok := w.(http.Flusher); ok {
        wf.Flush()
        log.Info("We have flushed successfully")
    }
}
```

The code above thinks that it has flushed something, but really it did not in a case when w is negroni.ResponseWriter.

The same task of wrapping http.ResponseWriter to save info (status, response size and etc.) is solved in promhttp package of client_golang. Their [tricky implementation](https://github.com/prometheus/client_golang/blob/main/prometheus/promhttp/delegator.go) lacks the problem mentioned above.

Sorry, that I don't have links to some real and not toy code, that relies on `wf, ok := w.(http.Flusher); ok` kind of check. But I'm pretty sure that someone may be affected by the problem.

### PR DESCRIPTION
**Pull request by jizhuozhi**
Fix #265

### CODE REVIEW

File: response_writer.go [hunk 1]
```diff
@@ -13,7 +10,7 @@ import (
 // if the functionality calls for it.
 type ResponseWriter interface {
 	http.ResponseWriter
-	http.Flusher
```
Comment (dearchap) [inline response_writer.go:16]: never a good idea to remove a method from a public interface

**Review by dearchap — APPROVED** [review]

File: response_writer.go (hunk 1 above)
Comment (jizhuozhi) [reply response_writer.go:16]: Based on my experience, this interface implementing would never be used directly as `w.Flush()` but likes below
```go
if f, ok := w.(http.Flusher); ok {
     f.Flush()
}
```
This change does not affect the contract, and follows the issue #265

Comment (jszwedko) [reply response_writer.go:16]: I agree that calling code typically looks like that and so the impact is likely to be minimal, but I still think this interface change is a breaking change and so we should put it in version 3 when releasing.

File: response_writer_test.go [hunk 2]
```diff
@@ -156,11 +156,8 @@ func TestResponseWriterHijack(t *testing.T) {
 func TestResponseWriteHijackNotOK(t *testing.T) {
 	hijackable := new(http.ResponseWriter)
 	rw := NewResponseWriter(*hijackable)
-	hijacker, ok := rw.(http.Hijacker)
```
Comment (jszwedko) [inline response_writer_test.go:159]: This test seems to be inverted now. Doesn't that indicate that we no longer implement hijackable when the wrapped writer does?

File: response_writer.go [hunk 3]
```diff
@@ -96,39 +89,8 @@ func (rw *responseWriter) Before(before func(ResponseWriter)) {
 	rw.beforeFuncs = append(rw.beforeFuncs, before)
 }
 
-func (rw *responseWriter) Hijack() (net.Conn, *bufio.ReadWriter, error) {
```
Comment (jszwedko) [inline response_writer.go:99]: Same here, I think this interface break should prompt us to bump the major version.

**Review by jszwedko — CHANGES_REQUESTED** [review]
Thanks for this @jizhuozhi !

I left a couple of questions, but mostly looks good.

File: response_writer_test.go (hunk 2 above)
Comment (jszwedko) [reply response_writer_test.go:159]: Ah, never mind, I misread this. I see this is testing that it returned an error, previously, if the wrapped response writer didn't implement `Hijacker`.

**Review by jszwedko — APPROVED** [review]
Realized I misread the below test. This looks good to me, thanks again @jizhuozhi ! I'll merge in. We'll just want to tag it as v3.0.0 when releasing.

**Pull request merged**